### DISCUSSION TITLE
How to close the current page by button click?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1542] [Q&A] [answered]
### Question

Hi All,

I am trying to create a button, which when pressed, will close the current page on the browser.

Recently in my another question #1529 , I learnt how to use the ```run_javascript``` feature.
So I applied that and used a function like this:
```python
async def close_page():
    await ui.run_javascript('window.close()')
```

So this works, the page does close. But I am getting some warning messages in the console. Is this normal to happen? Or is there a correct way to do this?

I am sharing the console warning message below for reference:

```python
Task exception was never retrieved
future: <Task finished name='Task-50' coro=<AsyncServer._handle_event_internal() done, defined at C:\apps\test\hivespace\Lib\site-packages\socketio\asyncio_server.py:522> exception=KeyError('result')>
Traceback (most recent call last):
  File "C:\apps\test\hivespace\Lib\site-packages\socketio\asyncio_server.py", line 524, in _handle_event_internal
    r = await server._trigger_event(data[0], namespace, sid, *data[1:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\apps\test\hivespace\Lib\site-packages\socketio\asyncio_server.py", line 562, in _trigger_event
    ret = handler(*args)
          ^^^^^^^^^^^^^^
  File "C:\apps\test\hivespace\Lib\site-packages\nicegui\nicegui.py", line 187, in on_javascript_response
    handle_javascript_response(client, msg)
  File "C:\apps\test\hivespace\Lib\site-packages\nicegui\nicegui.py", line 191, in handle_javascript_response
    client.waiting_javascript_commands[msg['request_id']] = msg['result']
                                                            ~~~^^^^^^^^^^
KeyError: 'result'
JavaScript did not respond in time
Traceback (most recent call last):
  File "C:\apps\test\hivespace\Lib\site-packages\nicegui\events.py", line 388, in wait_for_result
    await result
  File "C:\apps\test\api_tests\code_tests.py", line 38, in close_page
    await ui.run_javascript('window.close()')
  File "C:\apps\test\hivespace\Lib\site-packages\nicegui\functions\javascript.py", line 27, in run_javascript
    return await client.run_javascript(code, respond=respond, timeout=timeout, check_interval=check_interval)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\apps\test\hivespace\Lib\site-packages\nicegui\client.py", line 145, in run_javascript
    raise TimeoutError('JavaScript did not respond in time')
TimeoutError: JavaScript did not respond in time
```

**Comment by natankeddem** [comment, marked as answer]
The `run_javascript` method not only executes the code, it also allows for data to be returned from the page. I am guessing the return hooks don't like that you closed the page before they could do their thing. This seems to keep it from trying to wait on data return:
```py
from nicegui import ui


async def close_page():
    await ui.run_javascript('window.close()', respond=False)
ui.button('Button', on_click=lambda: close_page())

ui.run()
```

**Reply by Anindya088** [reply]
Hi @natankeddem ,

Thanks a lot for your help. I just tried out and it works perfectly, now I am not getting any warning any more on the console :)

Can I please ask for your help on a related question? While testing on this thing, I realized the code is working fine in Chrome but not on Edge. So is there any browser specific thing I need to understand or learn to make it work on Edge?

I am trying to put this in a dialog box inside a button, so sharing my code for your reference here:
```python
ui.button('Show dialog', on_click=lambda: show_dialog())

with ui.dialog().props('persistent') as dialog:
    with ui.card():
        ui.label('Dialog box to close page')
        with ui.card_actions().props('align="right"').classes('w-[100%]'):
            ui.button('Go to other page').props('flat')
            ui.button('Close this page', on_click=lambda: close_page()).props('flat')

async def close_page():
    await ui.run_javascript('window.close()', respond=False)

def show_dialog():
    dialog.open()
```

**Reply by falkoschindler** [reply]
We're thinking about simplifying the `run_javascript` function in version 1.4 by removing the `respond` parameter (https://github.com/zauberzeug/nicegui/discussions/1471). The function should simply work - independent of if it's awaited or not.

**Comment by natankeddem** [comment]
To be honest I am not a Javascript expert. It does appear what you are trying to do isn't super straight forward in general. Maybe this link will help?

https://stackoverflow.com/questions/25937212/window-close-doesnt-work-scripts-may-close-only-the-windows-that-were-opene

**Reply by python-and-novella** [reply]
For nicegui 1.4.*,we should do like this:
```python3
@ui.page('/',title='index')
def index():
    ui.button('close',on_click=lambda : ui.run_javascript('window.open(location.href, "_self", "");window.close()'))
```
